sync/src/snapshot/mod.rs

```diff
@@ -84,18 +98,86 @@ impl Service {
 
 fn snapshot(db: &dyn HashDB, block_hash: BlockHash, chunk_root: H256, root_dir: &str) -> Result<(), SnapshotError> {
     let snapshot_dir = snapshot_dir(root_dir, &block_hash);
-    create_dir_all(snapshot_dir)?;
+    fs::create_dir_all(snapshot_dir)?;
 
     for chunk in Snapshot::from_hashdb(db, chunk_root) {
         let chunk_path = snapshot_path(root_dir, &block_hash, &chunk.root);
-        let chunk_file = File::create(chunk_path)?;
+        let chunk_file = fs::File::create(chunk_path)?;
         let compressor = ChunkCompressor::new(chunk_file);
         compressor.compress_chunk(&chunk)?;
     }
 
     Ok(())
 }
 
+fn cleanup_expired(client: &Client, root_dir: &str, expiration: u64) -> Result<(), SnapshotError> {
+    for entry in fs::read_dir(root_dir)? {
+        let entry = match entry {
+            Ok(entry) => entry,
+            Err(err) => {
+                cerror!(SYNC, "Snapshot cleanup can't retrieve entry. err: {}", err);
+                continue
+            }
+        };
+        let path = entry.path();
+
+        match entry.file_type().map(|x| x.is_dir()) {
+            Ok(true) => {}
+            Ok(false) => continue,
+            Err(err) => {
+                cerror!(SYNC, "Snapshot cleanup can't retrieve file info: {}, err: {}", path.to_string_lossy(), err);
+                continue
+            }
+        }
+
+        let name = match path.file_name().expect("Directories always have file name").to_str() {
+            Some(n) => n,
+            None => continue,
+        };
+        let hash = match H256::from_str(name) {
+            Ok(h) => BlockHash::from(h),
+            Err(_) => continue,
+        };
+        let number = if let Some(number) = client.block_number(&BlockId::Hash(hash)) {
+            number
+        } else {
+            cerror!(SYNC, "Snapshot cleanup can't retrieve block number for block_hash: {}", hash);
+            continue
+        };
+
+        if number + expiration < client.best_block_header().number() {
+            do_cleanup(root_dir, hash)
+        }
+    }
+    Ok(())
+}
+
+/// Remove all files in `root_dir/block_hash`
+fn do_cleanup(root_dir: &str, block_hash: BlockHash) {
```

Comment:
How about passing `path` directly instead of `root_dir` and `block_hash`?

Reply:
Then the interface will be too generic.

Reply:
I think it doesn't matter since it's a private method.
In addition, what it tries to do is just removing a directory with a certain path, so I think passing `path` is more appropriate.